sources/osg/Uniform.js

```diff
@@ -26,395 +27,217 @@ Uniform.isUniform = function ( obj ) {
+        this._data[ 3 ] = f[ 3 ];
+        this._data[ 4 ] = f[ 4 ];
+        this._data[ 5 ] = f[ 5 ];
+        this._data[ 6 ] = f[ 6 ];
+        this._data[ 7 ] = f[ 7 ];
+        this._data[ 8 ] = f[ 8 ];
     },
-    _updateFloat16: function ( f ) {
-        this[ 0 ] = f[ 0 ];
-        this[ 1 ] = f[ 1 ];
-        this[ 2 ] = f[ 2 ];
-        this[ 3 ] = f[ 3 ];
-        this[ 4 ] = f[ 4 ];
-        this[ 5 ] = f[ 5 ];
-        this[ 6 ] = f[ 6 ];
-        this[ 7 ] = f[ 7 ];
-        this[ 8 ] = f[ 8 ];
-        this[ 9 ] = f[ 9 ];
-        this[ 10 ] = f[ 10 ];
-        this[ 11 ] = f[ 11 ];
-        this[ 12 ] = f[ 12 ];
-        this[ 13 ] = f[ 13 ];
-        this[ 14 ] = f[ 14 ];
-        this[ 15 ] = f[ 15 ];
+
+    setFloat16: function ( f ) {
+        this._data[ 0 ] = f[ 0 ];
+        this._data[ 1 ] = f[ 1 ];
+        this._data[ 2 ] = f[ 2 ];
+        this._data[ 3 ] = f[ 3 ];
+        this._data[ 4 ] = f[ 4 ];
+        this._data[ 5 ] = f[ 5 ];
+        this._data[ 6 ] = f[ 6 ];
+        this._data[ 7 ] = f[ 7 ];
+        this._data[ 8 ] = f[ 8 ];
+        this._data[ 9 ] = f[ 9 ];
+        this._data[ 10 ] = f[ 10 ];
+        this._data[ 11 ] = f[ 11 ];
+        this._data[ 12 ] = f[ 12 ];
+        this._data[ 13 ] = f[ 13 ];
+        this._data[ 14 ] = f[ 14 ];
+        this._data[ 15 ] = f[ 15 ];
     }
 };
+Uniform.prototype.setVec2 = Uniform.prototype.setFloat2;
+Uniform.prototype.setVec3 = Uniform.prototype.setFloat3;
+Uniform.prototype.setVec4 = Uniform.prototype.setFloat4;
+Uniform.prototype.setMatrix4 = Uniform.prototype.setFloat16;
+Uniform.prototype.setMatrix3 = Uniform.prototype.setFloat9;
 
-Uniform.createFloat1 = function ( data, uniformName ) {
+
+var createUniformX = function ( data, uniformName, defaultConstructor, glSignature, type, isMatrix ) {
     var value = data;
     var name = uniformName;
     if ( name === undefined ) {
         name = value;
-        value = [ 0 ];
+        value = defaultConstructor();
     }
-    var uniform = new Uniform();
+    var uniform = new Uniform( name );
 
-    // if we aleardy have an array because called from array version then dont generate
-    // an array of array
-    if ( !Array.isArray( value ) )
+    // if argument is not an array or typed array, create one
+    if ( !Array.isArray( value ) && value.byteLength === undefined )
         value = [ value ];
 
-    uniform.data = value;
-    uniform.glCall = function ( gl, location, glData ) {
-        gl.uniform1fv( location, glData );
-    };
-    uniform.glData = new MACROUTILS.Float32Array( uniform.data );
-    uniform.update = Uniform.prototype._updateFloat1;
-    uniform.set = function ( value ) {
-        if ( value.length === undefined ) {
-            this.data[ 0 ] = value;
-        } else {
-            this.data = value;
-        }
-        this.dirty();
-    };
-
-    uniform.name = name;
-    uniform.type = 'float';
+    uniform.setArray( value );
+    uniform._glCall = glSignature;
```

Comment:
not sure but couldn't we store the function pointer instead ?
```
uniform._glCall = gl[glSignature].bind(gl);
```
So that we do avoid a hash search each call of the gl.uniform
(as said, not sure if any interest)